commit 4d1a261202d828efc84e3a84d16c30548f29f76d
Author: Stef Walter <stefw@gnome.org>
Date:   Wed Apr 11 12:12:57 2012 +0200

    If canon'ing principals, write ccache with updated default principal
    
     * When calling krb5_get_init_creds_keytab() with
       krb5_get_init_creds_opt_set_canonicalize() the credential
       principal can get updated.
     * Create the cache file with the correct default credential.
     * LDAP GSSAPI SASL would fail due to the mismatched credentials
       before this patch.
    
    https://bugzilla.redhat.com/show_bug.cgi?id=811518

--- a/src/providers/krb5/krb5_child.c
+++ b/src/providers/krb5/krb5_child.c
@@ -626,7 +626,8 @@
         return kerr;
     }
 
-    kerr = create_ccache_file(ctx, princ, ccname, &creds);
+    /* Use the updated principal in the creds in case canonicalized */
+    kerr = create_ccache_file(ctx, creds.client, ccname, &creds);
     if (kerr != 0) {
         KRB5_DEBUG(1, kerr);
         goto done;
@@ -683,7 +684,10 @@
         }
     }
 
-    kerr = create_ccache_file(kr->ctx, kr->princ, kr->ccname, kr->creds);
+    /* Use the updated principal in the creds in case canonicalized */
+    kerr = create_ccache_file(kr->ctx,
+                              kr->creds ? kr->creds->client : kr->princ,
+                              kr->ccname, kr->creds);
     if (kerr != 0) {
         KRB5_DEBUG(1, kerr);
         goto done;
--- a/src/providers/ldap/ldap_child.c
+++ b/src/providers/ldap/ldap_child.c
@@ -285,7 +285,8 @@
         goto done;
     }
 
-    krberr = krb5_cc_initialize(context, ccache, kprinc);
+    /* Use updated principal if changed due to canonicalization. */
+    krberr = krb5_cc_initialize(context, ccache, my_creds.client);
     if (krberr) {
         DEBUG(2, ("Failed to init ccache: %s\n",
                   sss_krb5_get_error_message(context, krberr)));
